artiq/coredevice/ad53xx.py

```diff
@@ -0,0 +1,328 @@
+""""RTIO driver for the Analog Devices AD53[67][0123] family of multi-channel
+Digital to Analog Converters.
+
+Output event replacement is not supported and issuing commands at the same
+time is an error.
+"""
+
+# Designed from the data sheets and somewhat after the linux kernel
+# iio driver.
+
+from artiq.language.core import (kernel, portable, delay_mu, delay, now_mu,
+                                 at_mu)
+from artiq.language.units import ns
+from artiq.coredevice import spi2 as spi
+
+SPI_AD53XX_CONFIG = (0*spi.SPI_OFFLINE | 1*spi.SPI_END |
+                     0*spi.SPI_INPUT | 0*spi.SPI_CS_POLARITY |
+                     0*spi.SPI_CLK_POLARITY | 1*spi.SPI_CLK_PHASE |
+                     0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)
+
+AD53XX_CMD_DATA = 3 << 22
+AD53XX_CMD_OFFSET = 2 << 22
+AD53XX_CMD_GAIN = 1 << 22
+AD53XX_CMD_SPECIAL = 0 << 22
+
+AD53XX_SPECIAL_NOP = 0 << 16
+AD53XX_SPECIAL_CONTROL = 1 << 16
+AD53XX_SPECIAL_OFS0 = 2 << 16
+AD53XX_SPECIAL_OFS1 = 3 << 16
+AD53XX_SPECIAL_READ = 5 << 16
+
+AD53XX_READ_X1A = 0X000 << 7
+AD53XX_READ_X1B = 0X040 << 7
+AD53XX_READ_OFFSET = 0X080 << 7
+AD53XX_READ_GAIN = 0X0C0 << 7
+AD53XX_READ_CONTROL = 0X101 << 7
+AD53XX_READ_OFS0 = 0X102 << 7
+AD53XX_READ_OFS1 = 0X103 << 7
+
+
+@portable
+def ad53xx_cmd_write_ch(channel, value, op):
+    """Returns the word that must be written to the DAC to set a DAC
+    channel register to a given value.
+
+    :param channel: DAC channel to write to (8 bits)
+    :param value: 16-bit value to write to the register
+    :param op: The channel register to write to, one of
+      :const:`AD53XX_CMD_DATA`, :const:`AD53XX_CMD_OFFSET` or
+      :const:`AD53XX_CMD_GAIN`.
+    :return: The 24-bit word to be written to the DAC, aligned as the 24 MSB of
+      a 32-bit integer, ready to be transferred directly by the SPI core.
+    """
+    return (op | ((channel & 0x3f) + 8) << 16 | (value & 0xffff)) << 8
+
+
+@portable
+def ad53xx_cmd_read_ch(channel, op):
+    """Returns the word that must be written to the DAC to read a given
+    DAC channel register.
+
+    :param channel: DAC channel to read (8 bits)
+    :param op: The channel register to read, one of
+      :const:`AD53XX_CMD_DATA`, :const:`AD53XX_CMD_OFFSET` or
+      :const:`AD53XX_CMD_GAIN`
+    :return: The 24-bit word to be written to the DAC, aligned as the 24 MSB of
+      a 32-bit integer, ready to be transferred directly by the SPI core.
+    """
+    return (AD53XX_CMD_SPECIAL | AD53XX_SPECIAL_READ | op |
+           (((channel & 0x3f) + 8) << 7)) << 8
+
+
+@portable
+def voltage_to_mu(voltage, offset_dacs=8192, vref=5.):
```

Comment:
I think that calling both units of time and voltage `mu` is going to cause confusion at some point.

Reply:
I take your point, but we currently use mu for all units even freq phi and amp for dds. So this is just being consistent (which is the least confusing thing to do imo).

Reply:
IMO `mu` is ok for an opaque quantity of machine units independent of what the quantity describes if the corresponding SI unit is always clear from the context and it's not used as a variable name (`mu` itself is not ok, but `voltage_{to_}_mu` is fine).